pagespeed/system/controller_manager.cc

```diff
@@ -0,0 +1,228 @@
+// Copyright 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: jefftk@google.com (Jeff Kaufman)
+
+#include "pagespeed/system/controller_manager.h"
+
+#include <signal.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#include <cerrno>
+#include <cstddef>
+#include <cstdlib>
+
+#include "base/logging.h"
+#include "pagespeed/kernel/base/basictypes.h"
+#include "pagespeed/kernel/base/string_util.h"
+#include "pagespeed/system/system_rewrite_options.h"
+
+namespace net_instaweb {
+
+
+ControllerManager::ControllerManager(
+    SystemRewriteDriverFactory* factory,
+    ThreadSystem* thread_system,
+    MessageHandler* handler) : factory_(factory),
+                               thread_system_(thread_system),
+                               handler_(handler) { }
+
+void ControllerManager::SetUpSignalHandlers() {
+  // We need to clear inherited signal handlers. There's no portable way to get
+  // a list of all possible signals, and they're not even guaranteed to be in
+  // order.  But NSIG is usually defined these days, and if it is then we just
+  // want ascending numbers up to to NSIG.
+  for (int i = 0; i < NSIG; i++) {
+    signal(i, SIG_DFL);
```

Comment:
Depending what we're doing in this thread, that's not necessarily the right thing to do. For the controller in particular, it would be better to use the non RPC variant than to deal with all the serialisation overhead.